src/server/array_func.c

```diff
@@ -572,28 +572,43 @@ subst_array_index(job *pjob, char *path)
  *
  * @param[in]	range - subjob index range
  * @param[in]	initialstate - job state
+ * @param[out]	pbserror - PBSError to return
  *
  * @return	ptr to table
  * @retval  NULL	- error
  */
-static struct ajtrkhd *mk_subjob_index_tbl(char *range, int initalstate)
+static struct ajtrkhd *mk_subjob_index_tbl(char *range, int initalstate, int *pbserror)
 {
 	int   ct;
-	int   i, j;
+	int   i, j, l;
 	int   x, y, z;
 	char *eptr;
 	struct ajtrkhd *t;
 	size_t  sz;
 
 	i = parse_subjob_index(range, &eptr, &x, &y, &z, &ct);
-	if (i != 0)
+	if (i != 0) {
+		*pbserror = PBSE_BADATVAL;
 		return NULL; /* parse error */
+	}
+
+	if (server.sv_attr[(int)SVR_ATR_maxarraysize].at_flags & ATR_VFLAG_SET)
+		l = server.sv_attr[(int)SVR_ATR_maxarraysize].at_val.at_long;
+	else
+		l = PBS_MAX_ARRAY_JOB_DFL;  /* default limit 10000 */
+
+	if (ct > l) {
+		*pbserror = PBSE_MaxArraySize;
+		return NULL; /* parse error */
+	}
 
 	sz = sizeof(struct ajtrkhd) + ((ct-1) * sizeof(struct ajtrk));
 	t = (struct ajtrkhd *)malloc(sz);
 
-	if (t == NULL)
+	if (t == NULL) {
+		*pbserror = PBSE_SYSTEM;
 		return NULL;
+	}
```

Comment:
we need to keep the count level information about the job array for performance reasons, but why do need to keep the following:

t->tkm_tbl[j].trk_index = i;
 		t->tkm_tbl[j].trk_status = initalstate;
 		t->tkm_tbl[j].trk_error  = 0;
 		t->tkm_tbl[j].trk_substate  = JOB_SUBSTATE_FINISHED;
 		t->tkm_tbl[j].trk_stgout  = -1;
 		t->tkm_tbl[j].trk_exitstat  = 0;

These are supposed to be available directly from the subjobs themselves (or in case they were never instantiated they would simply be the defaults). So why an array of duplicate data that now actually exists in memory?

Reply:
Let us evaluate the effort that would be required to remove the tracking table from memory. If it is small we can implement it now. That way, not only we will not use more memory unnecessarily, but we will have less of a sync problem. If the effort turns out to be large, we should file a bug and address that later.

Reply:
I agree we are having a duplication of data in memory and the array in the tracking table should be unnecessary.  

Regarding the estimation of effort needed:  Seeing in the code that tkm_tbl is used more than forty places and seems to be tightly woven around. Removing it with proper quality can easily be a man month effort.

Reply:
>remove the tracking table from memory

Before we decide to do this in a future RFE, we need to think on how to handle when history is disabled, as subjobs are purged after completion before the complete job array finishes.